modules/dnn/src/layers/topk2_layer.cpp

```diff
@@ -0,0 +1,250 @@
+                    cmp.minusOffset(offset);
+                }
+            }
+        };
+        parallel_for_(Range(0, static_cast<int>(outer)), worker);
+    }
+
+public:
+    void forward(InputArrayOfArrays inputs_arr,
+                 OutputArrayOfArrays outputs_arr,
+                 OutputArrayOfArrays internals_arr) CV_OVERRIDE
+    {
+        CV_TRACE_FUNCTION();
+        std::vector<Mat> inputs, outputs;
+        inputs_arr.getMatVector(inputs);
+        outputs_arr.getMatVector(outputs);
+
+        const auto &input = inputs.front();
+        auto &output_value = outputs.front();
+        auto &output_index = outputs.back();
+
+        // Normalize axis to handle negative values
+        int normalized_axis = normalize_axis(axis, input.dims);
+
+        if (dynamicK) {
+            CV_Assert(inputs.size() == 2);
+            int64_t kVal = inputs[1].at<int64_t>(0);
+            CV_CheckGT(kVal, 0, "TopK2: dynamic K must be > 0");
+            CV_CheckLT(kVal, input.size[normalized_axis] + 1, "TopK2: dynamic K is out of range");
+            K = static_cast<int>(kVal);
+
+            MatShape outShape = shape(input);
+            outShape[normalized_axis] = K;
+
+            auto kind = outputs_arr.kind();
+            if (kind == _InputArray::STD_VECTOR_MAT) {
+                std::vector<Mat>& outs = outputs_arr.getMatVecRef();
+                CV_Assert(outs.size() == 2);
+                outs[0].fit(outShape, input.type());
+                outs[1].fit(outShape, CV_64S);
+                output_value = outs[0];
+                output_index = outs[1];
+            } else if (kind == _InputArray::STD_VECTOR_UMAT) {
+                std::vector<UMat>& uouts = outputs_arr.getUMatVecRef();
+                CV_Assert(uouts.size() == 2);
+                uouts[0].fit(outShape, input.type());
+                uouts[1].fit(outShape, CV_64S);
+                output_value = uouts[0].getMat(ACCESS_WRITE);
+                output_index = uouts[1].getMat(ACCESS_WRITE);
+            } else {
+                // Fallback – recreate directly on provided Mat refs.
+                output_value.create(outShape, input.type());
+                output_index.create(outShape, CV_64S);
+            }
+        }
+
+        if (largest) {
+            switch (input.depth()) {
+                case CV_8U: FindTopK<ComparatorGreater<uint8_t>, uint8_t>(input, output_value, output_index, normalized_axis); break;
+                case CV_8S: FindTopK<ComparatorGreater<int8_t>, int8_t>(input, output_value, output_index, normalized_axis); break;
+                case CV_16U: FindTopK<ComparatorGreater<uint16_t>, uint16_t>(input, output_value, output_index, normalized_axis); break;
+                case CV_16S: FindTopK<ComparatorGreater<int16_t>, int16_t>(input, output_value, output_index, normalized_axis); break;
+                case CV_16F: FindTopK<ComparatorGreater<hfloat>, hfloat>(input, output_value, output_index, normalized_axis); break;
```

Comment:
1) add CV_16BF branch as well
2) sorting hfloat's on x86 and sorting bfloat's on any platform will be slow because there is no support in hardware for scalar arithmetics on such types. Add yet another parameter, WT (work type) to the template method, set it to the type itself for all types except for hfloat and bfloat. For hfloat and bfloat set it to float.
3) in the comment above I recommended you to cache sorted values into a temporary buffer. Use WT for elements of that buffer.